#rule NAME:
#    input: "path/to/inputfile", "path/to/other/inputfile"
#    output: "path/to/outputfile", "path/to/another/outputfile"
#    shell: "somecommand {input} {output}"


configfile: "config.yaml"

rule bwa_map:
    input:
        ref=config['genomic_ref'],
        fq1="/projects/wp3/nobackup/WGS/MGI/V300043112_WGS_20201023/fastq/{sample}_R1.fastq.gz",
        fq2="/projects/wp3/nobackup/WGS/MGI/V300043112_WGS_20201023/fastq/{sample}_R2.fastq.gz"
    output:
        "mapped_reads/{sample}.bam"
    shell:
        "bwa mem {input.ref} {input.fq1} {input.fq2} | samtools view -Sb - > {output}"


rule samtools_sort:
    input:
        "mapped_reads/{sample}.bam"
    output:
        "sorted_reads/{sample}.bam"
    shell:
        "samtools sort -T sorted_reads/{wildcards.sample}.bam "
        "-O bam {input} > {output}"


rule samtools_index:
    input:
        "sorted_reads/{sample}.bam"
    output:
        "sorted_reads/{sample}.bam.bai"
    shell:
        "samtools index {input}"


rule deepvariant:
    input:
        ref = config['genomic_ref'],
        reads = "sorted_reads/{sample}.bam"
    output:
        ovcf = 'deepvariant/{sample}.vcf.gz',
        ogvcf = 'deepvariant/{sample}.g.vcf.gz',
        ird = directory('intermediate/{sample}')
    threads:
        8
    singularity:
        config['images']['deepvariant']
    log:
        '{sample}/log.txt'
    shell:
        "/opt/deepvariant/bin/run_deepvariant \
        --model_type=WGS \
        --ref={input.ref} \
        --reads={input.reads} \
        --output_vcf={output.ovcf} \
        --output_gvcf={output.ogvcf} \
        --intermediate_results_dir {output.ird} \
        --num_shards={threads} > {log} 2>&1"


rule all:
    input:
        "deepvariant/A.g.vcf.gz"
