# Result files
Poirot produces a lot of intermediate and result files but only files defined in [`output_files.json`](https://github.com/clinical-genomics-uppsala/poirot_rd_wgx/blob/main/config/output_files.json) are kept, the rest are temporary and will be deleted when not needed by any other rules in the pipeline. Please use the `--no-temp` snakemake option when running the pipeline to retain all files generated by the pipeline, or edit the output_files.json to retain a retain the files required.


## Files
Successful execution of the pipeline result in the the following output files located in `Results/`-folder:

| File | Format |Description |
|---|---|---|
multiqc_DNA.html| html  | Aggregated QC values for entire sequence run, open in browser|
{sample}/{sample}_N.cram | cram | Sorted and deduplicated CRAM alignement file |
{sample}/{sample}_N.cram.crai | crai |  Index for the CRAM alignment file | 
{sample}/spring/{sample}_{type}_{flowcell}_{lane}_{barcode}.spring| spring (.tar archives) | Spring compressed files (lossless) are tar archives of the raw fastq files (.fastq.gz files)|
{sample}/{sample}_snv_indels.vcf.gz | vcf.gz | Compressed deepvariant VCF file  |
{sample}/{sample}_snv_indels.vcf.gz.tbi | tbi | Index for the compressed deepvariant VCF file  |
{sample}/{sample}_snv_indels.filtered.vcf.gz | vcf.gz | Compressed deepvariant VCF file containing rare variants (filtered on VEP annoatoon max_af > 10%)  |
{sample}/{sample}_snv_indels.filtered.vcf.gz.tbi | tbi | Index for the compressed deepvariant VCF file containing rare variants  |
{sample}/cnv_sv/{sample}.cnvpytor_filtered.vcf.gz | vcf.gz | Filtered CNVpytor VCF file |
{sample}/cnv_sv/{sample}.cnvpytor.vcf.gz | vcf.gz | Unfiltered CNVpytor VCF file |
{sample}/cnv_sv/{sample}.cnvpytor_filtered.aed| aed | Filtered CNVpytor calls in aed format (Affymetrix Extensible Data format, tab delimited) |
{sample}/cnv_sv/{sample}.cnvpytor.aed | aed | Unfiltered CNVpytor calls in aed format (Affymetrix Extensible Data format, tab delimited) |
{sample}/cnv_sv/{sample}.tiddit.vcf.gz| vcf.gz  | Compressed VCF of SV called by Tiddit |
{sample}/cnv_sv/{sample}.manta_diploidSV.vcf.gz | vcf.gz  | Compressed VCF for SV calls called by Manta |
{sample}/cnv_sv/{sample}.svdb_merged.vcf.gz | vcf.gz | Compressed VCF containg the merge of SV and CNV calls from Tiddit, Manta and CNVpytor  |
{sample}/cnv_sv/{sample}.svdb_merged.filtered.vcf.gz | vcf.gz | A filtered version of the compressed VCF containing the merge of SV and CNV calls from Tiddit, Manta and CNVpytor |
{sample}/{sample}.upd_regions.bed | bed | Bed file of UPD regions called from a Trio VCF. Sample in this case is the sample id of the proband in the trio.  |
{sample}/{sample}.contamination.html | html | Contamination report from haplocheck program |
{sample}/{sample}.expansionhunter_stranger.vcf.gz | vcf.gz | Compressed VCF with repeat expansions called by Expansion Hunter and annotated with stranger |
{sample}/{sample}.expansionhunter_stranger.bed | bed |  Bed file with the repeat expansions called by Expansion Hunter and annotated with stranger |
{sample}/expansionhunter_reviewer | png  | Directory of png image files generated by REViewer|
{sample}/{sample}.coverage_analysis.xlsx | excel  | Excel file summarising the depth of covearage in genes in various gene panels |
{sample}/SMNCopyNumberCaller/{sample}.smn_charts.pdf | pdf | SMNCopyNumberCaller charts visualising the CN calls and read depth|
{sample}/SMNCopyNumberCaller/{sample}.smn_caller.tsv | tsv | SMNCopyNumberCaller results |
{sample}/SMNCopyNumberCaller/{sample}.smn_caller.json | json | SMNCopyNumberCaller results containing additional information not found in the SMNCopyNumberCaller tsv results   |
{sample}/automap/{sample}.HomRegions.tsv | tsv |  Regions of homozygosity  detected by AutoMap |
{sample}/automap/{sample}.HomRegions.pdf | pdf  | Visualisation of ROH regions detected by AutoMap |



## MultiQC report
Poirot produces a **[MultiQC](https://github.com/ewels/MultiQC)**-report for the entire sequencing run to enable easier QC tracking. The report starts with a general statistics table showing the most important QC-values followed by additional QC data and diagrams. The entire MultiQC html-file is interactive and you can filter, highlight, hide or export data using the ToolBox at the right edge of the report.

<br />

The report is configured based on a MultiQC config file. 

/// details | Expand to view current MultiQC config.yaml
```yaml
{% include "includes/multiqc_config.yaml" %}
```
///

The report contain qc information derived from the following programs:



### General Statistics
The general statistics table are ordered based on the sample order in Illumina `SampleSheet.csv` file used during sequencing, this is done by renaming the samples in two steps using the script `extract_samples_info.py`. To toggle between "Sample Order" and "Sample Name" use the buttons just above General Stats header.


<br />

| Column Name | Origin | Comment | 
| --- | --- | --- |
| M Reads | [Samtools stats](http://www.htslib.org/doc/samtools-stats.html)  | Total number of reads in the input nam fole |
| % Mapped| [Samtools stats](http://www.htslib.org/doc/samtools-stats.html) | Percent reads mapped, anywhere in the reference (no design file used) |
| % Proper pairs| [Samtools stats](http://www.htslib.org/doc/samtools-stats.html) | |
| Average Quality | [Samtools stats](http://www.htslib.org/doc/samtools-stats.html) | Ratio between sum of base quality over total length. Only reads on target (`config[reference][design_bed]`) |
| Median | [Mosdepth](https://github.com/brentp/mosdepth) | Median Coverage over coding exon in design (`config[reference][exon_bed]`) |
| >= 30X | [Mosdepth](https://github.com/brentp/mosdepth) | Fraction of coding exons (`config[reference][exon_bed]`) with coverage over 30x |
| >=50X |[Mosdepth](https://github.com/brentp/mosdepth) | Fraction of coding exons (`config[reference][exon_bed]`) with coverage over 50x |
| Bases on Target | [Picard](https://broadinstitute.github.io/picard/) HSMetrics | Bases inside the capture design (`config[reference][design_intervals]`) |
| Fold80 |[Picard](https://broadinstitute.github.io/picard/) HSMetrics | The fold over-coverage necessary to raise 80% of bases in "non-zero-cvg" targets to the mean coverage level in those targets (`config[reference][design_intervals]`) |
| % Dups | [Picard](https://broadinstitute.github.io/picard/) DuplicationMetrics |  |
| Mean Insert Size | [Picard](https://broadinstitute.github.io/picard/) InsertSizeMetrics |  |
| Target Bases with zero coverage [%] | [Picard](https://broadinstitute.github.io/picard/) HSMetrics | Percent target (`config[reference][design_intervals]`) bases with 0 coverage |